### ISSUE TITLE
AttributeError incorrectly catched by EventDispatcher.dispatch_event

### ISSUE THREAD

**Issue by marcvigoanglada** [pyglet/pyglet#105]
The following is a [simple example](https://github.com/pyglet/pyglet/files/3955261/pyglethello.py.txt) of the error we detected. 

We expect calling HelloWorldWindow.on_draw to raise an AttributeError exception. Instead, no error is raised but in the example the label is not shown (in fact, the handler is executed only until the wrong attribute is accessed). 

We traced the error and concluded that EventDispatcher.dispatch_event method incorrectly catches the AttributeError. This method should only catch the error raised by `gettatr.`

We propose to rewrite lines 418-427 in file pyglet/pyglet/event.py:

        # Check instance for an event handler
        try:
            if getattr(self, event_type)(*args):
                return EVENT_HANDLED
        except AttributeError:
            pass
        except TypeError as exception:
            self._raise_dispatch_exception(event_type, args, getattr(self, event_type), exception)
        else:
            invoked = True

like this:

        # Check instance for an event handler
        try:
            handler = getattr(self, event_type)
            try:
                if handler(*args):
                    return EVENT_HANDLED
            except TypeError as exception:
                self._raise_dispatch_exception(event_type, args, getattr(self, event_type), exception)
            else:
                invoked = True
        except AttributeError:
            pass

**Comment by caffeinepills** [comment]
Related to: 
https://github.com/pyglet/pyglet/issues/87

Looks like it was reverted back to the original way for 1.4.8:
`Revert Event Dispatcher changes from v1.4.7 due to possible bug.`

I'm not sure what the bug was but maybe this is a better way of catching the error?

**Comment by benmoran56** [comment]
I reverted this because it was causing wierd issues with some code. I think there might be a hard to find race condition with class instantiation. 

The main issue is that we're relying on an AttributeError to determine if a class has a handler method or not. Maybe we need to rethink how this code works, and make a more robust solution.

**Comment by pvcraven** [comment]
I still get this issue in 1.4.9. Or one like it. The pass below eats a lot of errors:

```
        # Check instance for an event handler
        try:
            if getattr(self, event_type)(*args):
                return EVENT_HANDLED
        except AttributeError:
            pass
        except TypeError as exception:
            self._raise_dispatch_exception(event_type, args, getattr(self, event_type), exception)
        else:
            invoked = True```

**Comment by benmoran56** [comment]
For more information, the "fix" was reverted due to causing events not to fire at all in some cases. It didn't always happen, but I was able to replicate the behavior with some of our examples. It was a bit baffling, so I reverted the code with the intention of giving it a review/rewrite when I have time.

**Comment by benmoran56** [comment]
This is now fixed since pyglet 1.5.4. Please give it a try with the latest release. If you still have issues, please reopen this issue or comment for follow up. Thanks!
